### PR TITLE
rc2: vendor github.com/dgryski/go-rc2

### PR DESCRIPTION
**Pull request by dgryski**
This eliminates an external dependency and simplifies the addition of
go-pkcs12 to /x/crypto.

### CODE REVIEW

File: rc2/rc2.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
+// Package rc2 implements the RC2 cipher
+/*
+https://www.ietf.org/rfc/rfc2268.txt
+http://people.csail.mit.edu/rivest/pubs/KRRR98.pdf
+
+This code is licensed under the MIT license.
```
Comment (nathany) [inline rc2/rc2.go]: This license header may need to change.

File: rc2/rc2.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+// The rc2 block size in bytes
+const BlockSize = 8
+
+type rc2Cipher struct {
+	k [64]uint16
+}
+
+// New returns a new rc2 cipher with the given key and effective key length t1
+func New(key []byte, t1 int) (cipher.Block, error) {
+	// TODO(dgryski): error checking for key length
```
Comment (nathany) [inline rc2/rc2.go]: TODO: when would you like to address this?

File: rc2/rc2.go (hunk 1 above)
Comment (dgryski) [inline rc2/rc2.go]: The rest of go-pkcs12 is also under MIT. I can of course relicense it under the Go license when that change occurs.

File: rc2/rc2.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+*/
+package rc2
+
+import (
+	"crypto/cipher"
+	"encoding/binary"
+)
+
+// The rc2 block size in bytes
+const BlockSize = 8
```
Comment (nathany) [inline rc2/rc2.go]: should this be exported? You also have the BlockSize() method below

File: rc2/rc2.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+	}
+
+	r0 = r0 + c.k[r3&63]
+	r1 = r1 + c.k[r0&63]
+	r2 = r2 + c.k[r1&63]
+	r3 = r3 + c.k[r2&63]
+
+	for j <= 40 {
+
```
Comment (nathany) [inline rc2/rc2.go]: :lipstick: extra blank lines are inconsistent in these three loops

File: rc2/rc2.go (hunk 3 above)
Comment (dgryski) [inline rc2/rc2.go]: crypto/aes exports BlockSize too.

File: rc2/rc2.go (hunk 4 above)
Comment (dgryski) [inline rc2/rc2.go]: I've cleaned those up and will push -f when this review is done.

File: rc2/rc2.go (hunk 2 above)
Comment (dgryski) [inline rc2/rc2.go]: I'll fix this now.

File: rc2/rc2.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+	r0 := binary.LittleEndian.Uint16(src[0:])
+	r1 := binary.LittleEndian.Uint16(src[2:])
+	r2 := binary.LittleEndian.Uint16(src[4:])
+	r3 := binary.LittleEndian.Uint16(src[6:])
+
+	var j int
+
+	for j <= 16 {
+		// mix r0
```
Comment (nathany) [inline rc2/rc2.go]: As far as I can tell, the j <= 16, j <= 40, and j <=60 loops all perform the same operations. Is that true?

If you extract a function that rakes pointers to r0-r3, is the performance still good? (will Go inline it?)

Is there some other way to perform &63 on iterations 17 and 41 that is efficient while reducing duplicate code?

File: rc2/rc2_test.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+import (
+	"bytes"
+	"encoding/hex"
+	"testing"
+)
+
+func TestEncryptDecrypt(t *testing.T) {
+
+	// TODO(dgryski): add the rest of the test vectors from the RFC
```
Comment (nathany) [inline rc2/rc2_test.go]: TODO: when would you like to address this?

File: rc2/rc2.go [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+		// unmix r1
+		r1 = rotl16(r1, 16-2)
+		r1 = r1 - c.k[j] - (r0 & r3) - ((^r0) & r2)
+		j--
+
+		// unmix r0
+		r0 = rotl16(r0, 16-1)
+		r0 = r0 - c.k[j] - (r3 & r2) - ((^r3) & r1)
+		j--
+
```
Comment (nathany) [inline rc2/rc2.go]: same comments as for Encrypt

File: rc2/rc2_test.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+			129,
+		},
+	}
+
+	for _, tt := range tests {
+		k, _ := hex.DecodeString(tt.key)
+		p, _ := hex.DecodeString(tt.plain)
+		c, _ := hex.DecodeString(tt.cipher)
+
+		b, _ := New(k, tt.t1)
```
Comment (nathany) [inline rc2/rc2_test.go]: would it be worth adding err != nil checks on these four lines? t.Fatalf

File: rc2/rc2.go (hunk 3 above)
Comment (nathany) [inline rc2/rc2.go]: :+1:

**Comment by nathany** [comment]
Thanks @dgryski. I'm a crypto novice, so I'm commenting on surface level code style things. I'll let the fine folks at Microsoft do a proper review.

File: rc2/rc2.go (hunk 5 above)
Comment (dgryski) [inline rc2/rc2.go]: This was initially exactly as you described.  I changed it in https://github.com/dgryski/go-rc2/commit/b3709a2071e1df24564fc378c021df57a055a65d

Comment (nathany) [inline rc2/rc2.go]: :+1:

Fortunately this code is unlikely to change, though maybe a comment around line 90 would be warranted? (repeating this, performance that)

**Comment by nathany** [comment]
One last thing, I think all exported identifiers should have comments (golint). Comments on internal identifiers doesn't hurt either.

File: rc2/rc2_test.go (hunk 6 above)
Comment (dgryski) [inline rc2/rc2_test.go]: Skimming the RFC, it appears I already have.

Comment (nathany) [inline rc2/rc2_test.go]: :wink:

**Comment by paulmey** [comment]
The cipher is functioning fine on the pkcs12 test cases and we'll get the real crypto review when we try to contribute it to x/crypto, where the experts are.
Godoc comments on exported identifiers are a must-have for contributing to Go. Otherwise, LGTM.

**Comment by dgryski** [comment]
All exported identifiers have godoc.

**Comment by dgryski** [comment]
Today I received my Microsoft CLA approval.

**Comment by paulmey** [comment]
Awesome, thanks @dgryski

**Pull request merged**